solutions/haskell/06-set_get/code/app/Main.hs

```diff
@@ -0,0 +1,184 @@
+             | Echo Message
+             | Set Key Value
+             | Get Key
+data ApplicationError = UnknownCommand
+data ProgConfigs = ProgConfigs {
+    recvBytes :: Int }
+data CmdNames = CmdNames {
+    pingN :: Text,
+    echoN :: Text,
+    setN :: Text,
+    getN :: Text }
+data RedisSpecs = RedisSpecs {
+    port :: String,
+    pingDefault :: Response,
+    unknownCmd :: Response,
+    setSuccess :: Response,
+    nilString :: Response,
+    bulkStringId :: ByteString,
+    arrayId :: ByteString,
+    simpleStringId :: ByteString }
+
+main :: IO ()
+main = do
+    putStrLn $ "\r\n>>> Redis server listening on port " ++ port redisSpecs ++ " <<<"
+    redisDB <- setupDB
+    serve HostAny (port redisSpecs) $ \(socket, _address) -> do
+        putStrLn $ "successfully connected client: " ++ show _address
+        _ <- forever $ do
+            request <- recv socket $ recvBytes progConfigs
+            response <- do
+                case parseRequest request of
+                    Left _ -> return $ unknownCmd redisSpecs
+                    Right cmd -> exec cmd redisDB
+            _ <- send socket (encodeRESP response)
+
+            -- debug database
+            out <- readTVarIO redisDB
+            putStrLn $ "\r\n***\r\nRedis DB content:\r\n"++ showTree out ++ "***\r\n"
+        putStrLn $ "disconnected client: " ++ show _address
+
+progConfigs :: ProgConfigs
+progConfigs = ProgConfigs {
+                recvBytes  = 2048 }
+
+cmdNames :: CmdNames
+cmdNames = CmdNames {
+                pingN = "ping",
+                echoN = "echo",
+                setN  = "set",
+                getN  = "get" }
+
+redisSpecs :: RedisSpecs
+redisSpecs = RedisSpecs {
+                port           = "6379",
+                pingDefault    = "PONG",
+                unknownCmd     = "-ERR Unknown Command",
+                setSuccess     = "OK",
+                nilString      = "(nil)",
+                bulkStringId   = "$",
+                arrayId        = "*",
+                simpleStringId = "+" }
+
+setupDB :: IO (TVar DB)
+setupDB = newTVarIO empty
+
+encodeRESP :: Response -> Response
+encodeRESP s = B.concat ["+", s, "\r\n"]
+
+exec :: Command -> TVar DB -> IO Response
+exec Ping _ = return "PONG"
+exec (Echo msg) _ = return msg
+exec (Set key value) db = set key value db
+exec (Get key) db = get key db
+
+parseRequest :: Request -> Either ApplicationError Command
+parseRequest req = case parseResult of
+                       Left _    -> Left UnknownCommand
+                       Right cmd -> Right cmd
+                   where parseResult = parse parseToCommand "" req
+
+parseToCommand :: Parser Command
+parseToCommand = try parseEcho
+             <|> try parsePing
+             <|> try parseSet
+             <|> try parseGet
+
+cmpIgnoreCase :: Text -> Text -> Bool
+cmpIgnoreCase a b = toLower a == toLower b
+
+-- some tools escape backslashes
+crlfAlt :: Parser (Tokens ByteString)
+crlfAlt = "\\r\\n" <|> crlf
```

Comment:
@beplay is this actually part of the official Redis spec? i.e. would the official Redis server behave the same? If so, safe to keep in - if not, we might want to remove it.

Reply:
@rohitpaulk The reason why we added this is that if you use netcat (i.e. `nc -v localhost 6379`) to connect to the Redis server you implement in Haskell, it allows you to simulate the exact RESP format the Redis CLI would send. This means, that you can send Redis Arrays `*1\r\n$4\r\nping\r\n` or a Redis Simple String `+ping\r\n`to the server and see if it matches, or at which exact position the parsing fails.

However, when using netcat, backslashes are automatically escaped. With escaped backslashes, the input turns into `+ping\\r\\n`, which does not match with just the in-built `crlf`, hence the alternative crlf function that handles this case.

If you use the official redis-cli to send commands, `crlf` is just fine, however, you can only send `ping` rather than `+ping\r\n`.

If you think that this way of testing the implementation locally is not something your users would do, then happy to remove it. But it took me a while to figure out why that input always failed, so I thought it may help others, too..